File: frontend/src/lib/money.ts
/**
 * VND money formatting + parsing helpers.
 *
 * VND has no fractional unit; all amounts are integer values in `đồng`
 * (the smallest unit). See architecture.md §Implementation Patterns
 * "JSON & data formats — Money" and EXPERIENCE.md §Voice and Tone
 * for the voice rules these helpers enforce.
 */

const THOUSAND = 1_000;
const MILLION = 1_000_000;

export type MoneyMode = 'full' | 'compact';

/**
 * Render a VND amount as a Vietnamese-norm string.
 *
 * - `full` (default): dot-separated thousands + `đ` suffix → `"120.000đ"`.
 * - `compact`: bucketed to `k` / `tr` with one decimal max, no trailing zero → `"120k"`, `"1.2tr"`, `"2tr"`.
 *
 * Negative amounts are formatted with a leading `-`; in practice the app
 * never displays negatives (discounts are subtracted at split time).
 */
export function formatMoney(amount: number, mode: MoneyMode = 'full'): string {
  if (!Number.isFinite(amount)) return '';
  const sign = amount < 0 ? '-' : '';
  const abs = Math.abs(amount);

  if (mode === 'compact') {
    if (abs >= MILLION) return sign + compactUnit(abs / MILLION, 'tr');
    if (abs >= THOUSAND) return sign + compactUnit(abs / THOUSAND, 'k');
    return sign + String(abs);
  }

  // 'full' mode — Vietnamese norm uses dot as the thousands separator.
  // Intl.NumberFormat with 'vi-VN' locale uses '.' for thousands.
  return sign + abs.toLocaleString('vi-VN') + 'đ';
}

function compactUnit(value: number, suffix: string): string {
  // One decimal place; trim trailing ".0".
  const rounded = Math.floor(value * 10) / 10;
  const text = rounded.toFixed(1).replace(/\.0$/, '');
  return text + suffix;
}

/**
 * Parse a Vietnamese-form money string back to its integer VND value.
 *
 * Accepts (case-insensitive, whitespace-tolerant):
 * - Plain integer: `"120000"`
 * - Dot-thousands: `"120.000"`, `"1.234.567"`
 * - Comma-thousands: `"120,000"`
 * - Compact `k`/`nghìn`: `"120k"`, `"120 nghìn"`, `"1.5k"`
 * - Compact `tr`/`triệu`: `"1tr"`, `"1.5 triệu"`
 * - Trailing `đ`/`đồng`/`vnđ` (stripped)
 *
 * Throws on empty, garbage, or negative input.
 */
export function parseMoney(input: string): number {
  if (typeof input !== 'string') throw new Error('parseMoney: not a string');
  let s = input.trim().toLowerCase();
  if (!s) throw new Error('parseMoney: empty input');

  // Strip currency suffixes.
  s = s.replace(/\s*(đồng|vnđ|vnd|đ)\s*$/u, '');

  // Compact triệu / tr — must check before generic 'tr' since 'triệu' contains 't' followed by other chars.
  let m = s.match(/^([\d.,]+)\s*(triệu|tr)$/u);
  if (m) {
    const n = parseFloat(m[1].replace(/,/g, '.'));
    if (!Number.isFinite(n)) throw new Error(`parseMoney: not a number "${input}"`);
    return rejectNegative(Math.round(n * MILLION), input);
  }

  // Compact nghìn / k.
  m = s.match(/^([\d.,]+)\s*(nghìn|nghin|k)$/u);
  if (m) {
    const n = parseFloat(m[1].replace(/,/g, '.'));
    if (!Number.isFinite(n)) throw new Error(`parseMoney: not a number "${input}"`);
    return rejectNegative(Math.round(n * THOUSAND), input);
  }

  // Pure number with thousands separators (dot or comma).
  // Vietnamese norm: dot = thousands. We strip all dots and commas.
  if (/^-?[\d.,]+$/.test(s)) {
    const stripped = s.replace(/[.,]/g, '');
    const n = parseInt(stripped, 10);
    if (!Number.isFinite(n)) throw new Error(`parseMoney: not a number "${input}"`);
    return rejectNegative(n, input);
  }

  throw new Error(`parseMoney: cannot parse "${input}"`);
}

function rejectNegative(n: number, original: string): number {
  if (n < 0) throw new Error(`parseMoney: negative not allowed "${original}"`);
  return n;
}

/**
 * Render an amount as Vietnamese spoken form for screen-reader `aria-label`.
 *
 * Matches the COMPACT display's precision (NOT the underlying value's full
 * digit count) so the audio narration aligns with what the user sees on
 * the tile.
 *
 * Examples:
 * - `120000` → `"120 nghìn đồng"`
 * - `1234567` → `"1 triệu 200 nghìn đồng"` (rounds to 1.2tr precision)
 * - `2000000` → `"2 triệu đồng"`
 * - `1500000` → `"1 triệu 500 nghìn đồng"`
 * - `0` → `"0 đồng"`
 */
export function moneySpoken(amount: number): string {
  if (!Number.isFinite(amount) || amount < 0) return '';
  if (amount === 0) return '0 đồng';

  if (amount >= MILLION) {
    // Compact "tr" is value / 1_000_000, rounded down to 1 decimal place.
    // The decimal tenth maps to "× 100 nghìn" (so 1.2tr = 1 triệu 200 nghìn).
    const tenths = Math.floor(amount / 100_000); // 12 for 1234567
    const wholeMillions = Math.floor(tenths / 10); // 1
    const tenthDigit = tenths % 10; // 2
    const parts: string[] = [`${wholeMillions} triệu`];
    if (tenthDigit > 0) parts.push(`${tenthDigit * 100} nghìn`);
    return parts.join(' ') + ' đồng';
  }

  if (amount >= THOUSAND) {
    // Round to nearest thousand (matches the "k" compact display).
    const k = Math.round(amount / THOUSAND);
    return `${k} nghìn đồng`;
  }

  return `${amount} đồng`;
}

File: backend/internal/money/money.go
// Package money provides VND formatting + parsing helpers.
//
// Mirrors frontend/src/lib/money.ts in behavior — keep them in sync when
// either side changes (Story 1.3 establishes the contract).
//
// VND has no fractional unit; all amounts are integer values in `đồng`
// (the smallest unit). See architecture.md §Implementation Patterns
// "JSON & data formats — Money".
package money

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

const (
	thousand int64 = 1_000
	million  int64 = 1_000_000
)

// Mode controls how Format renders an amount.
type Mode int

const (
	// Full renders with dot-separated thousands and a `đ` suffix (Vietnamese norm).
	Full Mode = iota
	// Compact renders bucketed to k / tr with one decimal, no trailing `.0`.
	Compact
)

// Format an amount as a Vietnamese-norm money string.
//
// See Mode for behavior. Negative amounts are formatted with a leading `-`;
// the app does not normally display negatives.
func Format(amount int64, mode Mode) string {
	sign := ""
	abs := amount
	if abs < 0 {
		sign = "-"
		abs = -abs
	}

	if mode == Compact {
		switch {
		case abs >= million:
			return sign + compactUnit(float64(abs)/float64(million), "tr")
		case abs >= thousand:
			return sign + compactUnit(float64(abs)/float64(thousand), "k")
		default:
			return sign + strconv.FormatInt(abs, 10)
		}
	}

	// Full mode — insert a `.` every 3 digits from the right.
	digits := strconv.FormatInt(abs, 10)
	n := len(digits)
	var b strings.Builder
	for i, r := range digits {
		if i > 0 && (n-i)%3 == 0 {
			b.WriteByte('.')
		}
		b.WriteRune(r)
	}
	b.WriteString("đ")
	return sign + b.String()
}

// compactUnit formats a float as "<value><suffix>" with one decimal max
// and no trailing `.0`.
func compactUnit(value float64, suffix string) string {
	// Truncate (not round) to one decimal — matches the TS Math.floor behavior.
	truncated := float64(int64(value*10)) / 10
	text := strconv.FormatFloat(truncated, 'f', 1, 64)
	text = strings.TrimSuffix(text, ".0")
	return text + suffix
}

var (
	// Trailing currency markers we strip during parse.
	currencySuffixRe = regexp.MustCompile(`(?i)\s*(đồng|vnđ|vnd|đ)\s*$`)
	// Compact `triệu` / `tr` forms.
	compactMillionRe = regexp.MustCompile(`^([\d.,]+)\s*(triệu|tr)$`)
	// Compact `nghìn` / `nghin` / `k` forms.
	compactThousandRe = regexp.MustCompile(`^([\d.,]+)\s*(nghìn|nghin|k)$`)
	// Pure numeric (with optional dots / commas as thousands separators).
	plainNumericRe = regexp.MustCompile(`^[\d.,]+$`)
)

// Parse a Vietnamese-form money string back to its integer VND value.
//
// See ParseExamples in the package-level docs for what inputs are accepted.
// Returns an error on empty, garbage, or negative input.
func Parse(input string) (int64, error) {
	s := strings.ToLower(strings.TrimSpace(input))
	if s == "" {
		return 0, fmt.Errorf("money.Parse: empty input")
	}

	// Strip currency suffix.
	s = currencySuffixRe.ReplaceAllString(s, "")
	s = strings.TrimSpace(s)

	// Compact triệu / tr — check before nghìn / k.
	if m := compactMillionRe.FindStringSubmatch(s); m != nil {
		f, err := parseDecimal(m[1])
		if err != nil {
			return 0, fmt.Errorf("money.Parse: not a number %q", input)
		}
		return rejectNegative(int64(f*float64(million)+0.5), input)
	}

	// Compact nghìn / k.
	if m := compactThousandRe.FindStringSubmatch(s); m != nil {
		f, err := parseDecimal(m[1])
		if err != nil {
			return 0, fmt.Errorf("money.Parse: not a number %q", input)
		}
		return rejectNegative(int64(f*float64(thousand)+0.5), input)
	}

	// Plain number with thousands separators.
	if plainNumericRe.MatchString(s) {
		stripped := strings.ReplaceAll(strings.ReplaceAll(s, ".", ""), ",", "")
		n, err := strconv.ParseInt(stripped, 10, 64)
		if err != nil {
			return 0, fmt.Errorf("money.Parse: not a number %q", input)
		}
		return rejectNegative(n, input)
	}

	return 0, fmt.Errorf("money.Parse: cannot parse %q", input)
}

// parseDecimal accepts "1.5" or "1,5" as 1.5.
func parseDecimal(s string) (float64, error) {
	s = strings.ReplaceAll(s, ",", ".")
	return strconv.ParseFloat(s, 64)
}

func rejectNegative(n int64, original string) (int64, error) {
	if n < 0 {
		return 0, fmt.Errorf("money.Parse: negative not allowed %q", original)
	}
	return n, nil
}

// Spoken renders an amount as Vietnamese spoken form for screen-reader
// `aria-label`. Matches the COMPACT display's precision (not the underlying
// value's full digits) so the audio aligns with the visible tile.
//
// Examples:
//
//	Spoken(120_000)   == "120 nghìn đồng"
//	Spoken(1_234_567) == "1 triệu 200 nghìn đồng"  (rounds to 1.2tr precision)
//	Spoken(2_000_000) == "2 triệu đồng"
//	Spoken(0)         == "0 đồng"
func Spoken(amount int64) string {
	if amount < 0 {
		return ""
	}
	if amount == 0 {
		return "0 đồng"
	}

	if amount >= million {
		tenths := amount / 100_000   // 12 for 1234567
		wholeMillions := tenths / 10 // 1
		tenthDigit := tenths % 10    // 2
		parts := []string{fmt.Sprintf("%d triệu", wholeMillions)}
		if tenthDigit > 0 {
			parts = append(parts, fmt.Sprintf("%d nghìn", tenthDigit*100))
		}
		return strings.Join(parts, " ") + " đồng"
	}

	if amount >= thousand {
		// Round to nearest thousand (matches the "k" compact display).
		k := (amount + 500) / thousand
		return fmt.Sprintf("%d nghìn đồng", k)
	}

	return fmt.Sprintf("%d đồng", amount)
}
